[SPARK-30564][SQL] Revert Block.length and use comment placeholders in HashAggregateExec

peter-toth opened the pull request:
### What changes were proposed in this pull request?
SPARK-21870 (cb0cddf#diff-06dc5de6163687b7810aa76e7e152a76R146-R149) caused significant performance regression in cases where the source code size is fairly large as `HashAggregateExec` uses `Block.length` to decide on splitting the code. The change in `length` makes sense as the comment and extra new lines shouldn't be taken into account when deciding on splitting, but the regular expression based approach is very slow and adds a big relative overhead to cases where the execution is quick (small number of rows).
This PR:
- restores `Block.length` to its original form
- places comments in `HashAggragateExec` with `CodegenContext.registerComment` so as to appear only when comments are enabled (`spark.sql.codegen.comments=true`)

Before this PR:
```
deeply nested struct field r/w:           Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
250 deep x 400 rows (read in-mem)                  1137           1143           8          0.1       11368.3       0.0X
```

After this PR:
```
deeply nested struct field r/w:           Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
250 deep x 400 rows (read in-mem)                   167            180           7          0.6        1674.3       0.1X
```
### Why are the changes needed?
To fix performance regression.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Existing UTs.

peter-toth committed:
[SPARK-30564][SQL] Use comment placeholders in HashAggregateExec

SparkQA commented:
**[Test build #120646 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120646/testReport)** for PR 28083 at commit [`e67f80d`](https://github.com/apache/spark/commit/e67f80dca6f3230a27bdc8dee83604476bc5037f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25348/
Test PASSed.

SparkQA commented:
**[Test build #120646 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120646/testReport)** for PR 28083 at commit [`e67f80d`](https://github.com/apache/spark/commit/e67f80dca6f3230a27bdc8dee83604476bc5037f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120646/
Test PASSed.

dongjoon-hyun commented:
cc @maropu and @cloud-fan

dongjoon-hyun commented:
Thank you for reporting, @peter-toth .
Can we have an example to verify?

maropu commented:
The change looks fine to me. Could you put the `WideSchemaBenchmark` performance numbers in the PR description before/after this PR?

maropu commented:
Any update? Still WIP?

peter-toth commented:
Yes still WIP, didn't have time to work on this unfortunately. I will try to sort this out during the weekend.

peter-toth commented:
> Yes still WIP, didn't have time to work on this unfortunately. I will try to sort this out during the weekend.

I still prefer the [other PR](https://github.com/apache/spark/pull/28041#issuecomment-612643258) instead of this one so let me leave this in WIP for now.

peter-toth commented:
I've added perf numbers as @maropu asked and removed WIP.

This PR can be an alternative to revert https://issues.apache.org/jira/browse/SPARK-21870 as a whole (https://github.com/apache/spark/pull/28041#issuecomment-613376751)

cc @kiszk @cloud-fan @dongjoon-hyun @rxin

kiszk commented:
Looks good to me

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #121300 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121300/testReport)** for PR 28083 at commit [`e67f80d`](https://github.com/apache/spark/commit/e67f80dca6f3230a27bdc8dee83604476bc5037f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25984/
Test PASSed.

SparkQA commented:
**[Test build #121300 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121300/testReport)** for PR 28083 at commit [`e67f80d`](https://github.com/apache/spark/commit/e67f80dca6f3230a27bdc8dee83604476bc5037f).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121300/
Test FAILed.

peter-toth commented:
retest this please

SparkQA commented:
**[Test build #121309 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121309/testReport)** for PR 28083 at commit [`e67f80d`](https://github.com/apache/spark/commit/e67f80dca6f3230a27bdc8dee83604476bc5037f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25993/
Test PASSed.

SparkQA commented:
**[Test build #121309 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121309/testReport)** for PR 28083 at commit [`e67f80d`](https://github.com/apache/spark/commit/e67f80dca6f3230a27bdc8dee83604476bc5037f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121309/
Test PASSed.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/javaCode.scala:
```diff
@@ -143,10 +143,7 @@ trait Block extends TreeNode[Block] with JavaCode {
   }
 
-  def length: Int = {
-    // Returns a code length without comments
```
Ur, one more comment; could you leave some comments here about why the current logic returns a length with comments?

peter-toth replied:
Ok, added.

maropu replied:
Thanks

peter-toth committed:
add comment to length

SparkQA commented:
**[Test build #121357 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121357/testReport)** for PR 28083 at commit [`84a3e1e`](https://github.com/apache/spark/commit/84a3e1ebfd68422c7f11f6c4c6dcdd968fdd8ace).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26041/
Test PASSed.

maropu commented:
The last Jenkins job passed and the last commit has only comments. So, I merged this to master/3.0. Thanks, all!

peter-toth commented:
Thanks all for the review.

SparkQA commented:
**[Test build #121357 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121357/testReport)** for PR 28083 at commit [`84a3e1e`](https://github.com/apache/spark/commit/84a3e1ebfd68422c7f11f6c4c6dcdd968fdd8ace).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121357/
Test PASSed.

dongjoon-hyun commented:
+1, late LGTM. Thank you all.
